### PR TITLE
fix terraform resource paths, use templatefile, update metal version

### PR DESCRIPTION
**Pull request by displague**
## Description

- Updates the EM provider to the latest
- Uses EM Metros
- Uses path.module in all local file references (this is a TF best practice and is required in order to consume this config as a module)
- replaces use of the deprecated template data source with the templatefile function
- adds the .terraform.lock.hcl file, another TF best practice
- use cloud-config (avoiding ssh provisioners)
- use `metal_port` resources instead of `metal_device_network_type` and `metal_vlan_port_attachment` resources (https://registry.terraform.io/providers/equinix/metal/latest/docs/guides/network_types). 

## Why is this needed

I needed these changes to pursue work in https://github.com/displague/terraform-metal-tinkerbell/pull/1
The project itself creates a Terraform module (which could be submitted to the Terraform registry) by thinly wrapping the sandbox project configs.

In the linked PR, I am attempted to use the Tinkerbell Terraform Provider to work with the created provisioner node.  I do not want to do anything but `terraform apply` to have an EM node execute a Tinkerbell workflow.

Fixes: #

## How Has This Been Tested?

Deployed via https://github.com/displague/terraform-metal-tinkerbell

## How are existing users impacted? What migration steps/scripts do we need?

Users will have to `terraform init -upgrade`, which is typical.  Users may also need a newer version of Terraform if they are running TF <0.12.

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: deploy/terraform/main.tf [hunk 1]
```diff
@@ -52,27 +52,27 @@ resource "null_resource" "tink_directory" {
   }
 
   provisioner "file" {
-    source      = "../../setup.sh"
+    source      = "${path.module}/../../setup.sh"
```
Comment (displague) [inline deploy/terraform/main.tf:55]: I am really suprised that this worked through the `../../`,  but it did:

https://github.com/displague/terraform-metal-tinkerbell/blob/main/main.tf#L2

```
root@tink-provisioner:~# ls -la /root/tink/
total 40
drwxr-xr-x 3 root root  4096 Aug 13 22:29 .
drwx------ 6 root root  4096 Aug 13 23:24 ..
-rwxr-xr-x 1 root root   771 Aug 13 22:29 current_versions.sh
drwxr-xr-x 6 root root  4096 Aug 13 22:31 deploy
-rwxr-xr-x 1 root root  2800 Aug 13 22:29 generate-env.sh
-rw-r--r-- 1 root root     6 Aug 13 22:29 .nat_interface
-rwxr-xr-x 1 root root 13594 Aug 13 22:29 setup.sh
```

**Comment by displague** [comment]
@jacobweinstock Yes, this is still needed. I've updated the PR (and the PR description) and verified that this terraform config works when run from the sandbox/deploy/terraform directory.

**Review by jacobweinstock — COMMENTED** [review]
Hey @displague, thanks for updating this. I'm not able to get this working. I'm seeing this below. Any ideas?

```bash
❯ terraform apply  
╷
│ Error: Invalid function argument
│ 
│   on main.tf line 102, in data "cloudinit_config" "setup":
│  102:       COMPOSE_ZIP = filebase64("${path.module}/compose.zip")
│     ├────────────────
│     │ path.module is "."
│ 
│ Invalid value for "path" parameter: no file exists at ./compose.zip; this function works only with files that are distributed as part of the configuration source code, so if this file will be created by a resource in this configuration you must instead obtain this result from an attribute of that resource.
╵

 ~/repos/tinkerbell/sandbox/deploy/terraform  pr/displague/96:terraform-paths                                                                                                                                                                                                                                                                                                                                      ▼  09:26:33 AM 
❯ terraform version
Terraform v1.0.4
on darwin_amd64
+ provider registry.terraform.io/equinix/metal v3.1.0
+ provider registry.terraform.io/hashicorp/archive v2.2.0
+ provider registry.terraform.io/hashicorp/cloudinit v2.2.0
+ provider registry.terraform.io/hashicorp/null v2.1.2
+ provider registry.terraform.io/hashicorp/template v2.1.2

Your version of Terraform is out of date! The latest version
is 1.0.6. You can update by downloading from https://www.terraform.io/downloads.html

```

**Comment by displague** [comment]
@jacobweinstock I'll push up a fix to work around some Terraform limitations here.  This works as-is on the second apply because compose.zip is created in the first failing pass.  I was looking at this problem specifically but missed it in my testing.

The work-around will involve a `locals` block which will insert a transitive phase in the planning, requiring the compose.zip to be created before attempting to use it. (the depends_on is not sufficient in this case)

**Comment by displague** [comment]
@jacobweinstock pushed. Thanks for taking a look!

**Comment by displague** [comment]
@jacobweinstock looks like that last push resulted in an empty /root/sandbox/compose dir. I'll have to take another look at how to work with or around these Terraform behaviors.

**Comment by jacobweinstock** [comment]
Hey @displague, anything I can do to help with this?

**Comment by displague** [comment]
@jacobweinstock was there talk of dismantling sandbox?

**Comment by jacobweinstock** [comment]
> @jacobweinstock was there talk of dismantling sandbox?

No, there wasn't. Why do you ask?

**Comment by nshalman** [comment]
Is there a desire for someone else to pick this up?

**Comment by displague** [comment]
@nshalman @jacobweinstock This PR is working as expected now. Please give it another look.

**Review by cprivite — APPROVED** [review]
We should update the documentation to reference metro instead of facility, but looks good otherwise. I'll be making at least some of those documentation changes in the next few days.

**Review by jacobweinstock — APPROVED** [review]
Hey @displague, really sorry for the terribly delayed response. This looks good. Only a doc update suggestion.

When I tested this I needed to change lines [45](https://github.com/tinkerbell/sandbox/blob/a2245d46f010dd8dd988bd75cb621c8478062aa1/docs/quickstarts/TERRAFORMEM.md?plain=1#L45) and [97](https://github.com/tinkerbell/sandbox/blob/a2245d46f010dd8dd988bd75cb621c8478062aa1/docs/quickstarts/TERRAFORMEM.md?plain=1#L97) in the terraform [quickstart doc](https://github.com/tinkerbell/sandbox/blob/main/docs/quickstarts/TERRAFORMEM.md).
`metal device reboot -i $(terraform show -json | jq -r '.values.root_module.resources[1].values.id')` -> `metal device reboot -i $(terraform show -json | jq -r '.values.root_module.resources[3].values.id')`

**Comment by displague** [comment]
@jacobweinstock Rebased and updated

File: deploy/terraform/main.tf [hunk 2]
```diff
@@ -23,86 +23,86 @@ provider "metal" {
...
-resource "metal_port_vlan_attachment" "provisioner" {
-  depends_on = [metal_device_network_type.tink_provisioner_network_type]
-  device_id  = metal_device.tink_provisioner.id
-  port_name  = "eth1"
-  vlan_vnid  = metal_vlan.provisioning_vlan.vxlan
+data "archive_file" "compose" {
+  type        = "zip"
+  source_dir  = "${path.module}/../compose"
+  output_path = "${path.module}/compose.zip"
 }
```
Comment (mmlb) [inline deploy/terraform/main.tf:84]: must we go with `zip` here to make this a module? Having to install unzip makes for a runtime failure (wan network issue...) vs using local-exec & tar the failure will happen early and solved "once".

Comment (displague) [reply deploy/terraform/main.tf:84]: `zip` is what is available today in the Terraform module. Taking advantage of what is available in Terraform reduces our dependence on local compression tools.

https://registry.terraform.io/providers/hashicorp/archive/latest/docs/data-sources/archive_file 

https://github.com/hashicorp/terraform-provider-archive/pull/29

Comment (displague) [reply deploy/terraform/main.tf:84]: We have other network dependencies that would fail with `zip`, be it package installs or Tink configuration.

**Pull request merged**